File: lightroom_tagger/core/identity_service/suggest_post.py
"""Post-next suggestions from coverage-eligible catalog gaps."""

from __future__ import annotations

import sqlite3
from typing import Any

from .mirror import build_mirror_scan, compute_signature_stats
from .percentiles import compute_image_peak_percentile_scores
from .ranking import _image_meta_map


def _peak_perspective(per_perspective: list[dict[str, Any]]) -> dict[str, Any]:
    return sorted(
        per_perspective,
        key=lambda p: (-float(p.get("percentile") or 0.0), str(p.get("perspective_slug") or "")),
    )[0]


def suggest_what_to_post_next(
    conn: sqlite3.Connection,
    *,
    limit: int,
    offset: int = 0,
    sort_by_date: str | None = None,
) -> dict[str, Any]:
    """Unposted, coverage-eligible catalog images with heuristic reasons (D-44–D-46).

    ``sort_by_date`` (``newest`` / ``oldest``) only controls the date tiebreaker;
    peak within-perspective percentile remains the primary sort key.
    """
    if sort_by_date is not None and sort_by_date not in ("newest", "oldest"):
        raise ValueError("sort_by_date must be 'newest' or 'oldest'")

    scan = build_mirror_scan(conn)
    items, peak_meta = compute_image_peak_percentile_scores(
        conn,
        include_ineligible=False,
        percentile_lookup=scan.percentile_lookup,
    )
    crowned = {s["perspective_slug"] for s in compute_signature_stats(scan) if s["crowned"]}
    keys = [str(i["image_key"]) for i in items if i.get("eligible")]
    img_meta = _image_meta_map(conn, keys)

    candidates_full: list[dict[str, Any]] = []
    for i in items:
        if not i.get("eligible"):
            continue
        k = str(i["image_key"])
        im = img_meta.get(k, {})
        if im.get("instagram_posted"):
            continue
        per = list(i.get("per_perspective") or [])
        peak_lens = _peak_perspective(per) if per else {}
        slug = str(peak_lens.get("perspective_slug") or "")
        candidates_full.append(
            {
                **i,
                **im,
                "peak_perspective_slug": slug,
                "peak_perspective_display_name": str(
                    peak_lens.get("display_name") or slug
                ),
                "is_signature": slug in crowned if slug else False,
            }
        )

    date_reverse = sort_by_date != "oldest"
    candidates_full.sort(key=lambda r: r["image_key"])
    candidates_full.sort(key=lambda r: r.get("date_taken") or "", reverse=date_reverse)
    candidates_full.sort(key=lambda r: r["peak_percentile"], reverse=True)

    unposted_peaks = [float(c["peak_percentile"]) for c in candidates_full]
    if unposted_peaks:
        sorted_peaks = sorted(unposted_peaks)
        idx = max(0, int(round(0.9 * (len(sorted_peaks) - 1))))
        p90 = sorted_peaks[idx]
    else:
        p90 = 1.0

    suggestions_meta: dict[str, Any] = {
        "weighting": peak_meta.get("weighting"),
        "ranking_key": peak_meta.get("ranking_key"),
        "min_perspectives_used": peak_meta.get("min_perspectives_used"),
        "coverage_rule": peak_meta.get("coverage_rule"),
        "high_score_rule": (
            "reason code high_score_unposted when peak_percentile >= p90 of eligible "
            "unposted images' peak within-perspective percentiles."
        ),
    }

    total_candidates = len(candidates_full)
    lim = max(0, limit)
    page = candidates_full[offset : offset + lim]

    out_candidates: list[dict[str, Any]] = []
    for cand in page:
        reasons: list[str] = []
        codes: list[str] = []
        peak = float(cand["peak_percentile"])
        lens_name = str(cand.get("peak_perspective_display_name") or "")
        is_sig = bool(cand.get("is_signature"))
        if peak >= p90:
            reasons.append(
                "Strong peak percentile among scored, unposted catalog images "
                f"(peak_percentile={peak:.4f})."
            )
            codes.append("high_score_unposted")

        if lens_name:
            if is_sig:
                reasons.append(
                    f"Peak lens is {lens_name}, one of your crowned signature techniques."
                )
            else:
                reasons.append(f"Peak lens is {lens_name}.")

        if not codes:
            reasons.append(
                "Unposted catalog image with sufficient perspective coverage; "
                "ranked by peak within-perspective percentile among eligible candidates."
            )
            codes.append("eligible_unposted")

        out_candidates.append(
            {
                "image_key": cand["image_key"],
                "filename": cand.get("filename", ""),
                "date_taken": cand.get("date_taken", ""),
                "rating": cand.get("rating", 0),
                "peak_percentile": cand["peak_percentile"],
                "peak_perspective_slug": cand.get("peak_perspective_slug", ""),
                "peak_perspective_display_name": cand.get("peak_perspective_display_name", ""),
                "is_signature": is_sig,
                "perspectives_covered": cand["perspectives_covered"],
                "per_perspective": cand.get("per_perspective", []),
                "reasons": reasons,
                "reason_codes": codes,
            }
        )

    empty_state: str | None = None
    if not out_candidates:
        empty_state = (
            "No unposted catalog images meet perspective coverage with current scores."
        )

    return {
        "candidates": out_candidates,
        "total": total_candidates,
        "meta": suggestions_meta,
        "empty_state": empty_state,
    }

File: apps/visualizer/backend/src/identity/suggest-post.ts
/**
 * Post-next suggestions from coverage-eligible catalog gaps.
 */
import type { Db } from '../db/connection.js';
import { computeImagePeakPercentileScores, type PeakPercentileItem } from './percentiles.js';
import { buildCatalogScoreIndex } from './score-index.js';
import { sortByRanking } from './sort.js';
import { imageMetaMap } from './ranking.js';
import { computeSignatureStats } from './signature.js';

/**
 * Round halves to even, not away from zero.
 *
 * Load-bearing for the p90 index: `round(0.9 * (n - 1))` at exactly 4.5 picks
 * index 4 under half-to-even but 5 under `Math.round`.
 */
export function roundHalfEven(value: number): number {
  const floor = Math.floor(value);
  const diff = value - floor;
  if (diff > 0.5) return floor + 1;
  if (diff < 0.5) return floor;
  return floor % 2 === 0 ? floor : floor + 1;
}

interface PerPerspectiveLike {
  perspective_slug?: string;
  display_name?: string;
  percentile?: number;
}

/** The image's strongest lens: highest percentile, then slug for determinism. */
function peakPerspective(perPerspective: readonly PerPerspectiveLike[]): PerPerspectiveLike {
  return [...perPerspective].sort((a, b) => {
    const byPct = (b.percentile ?? 0) - (a.percentile ?? 0);
    if (byPct !== 0) return byPct;
    const sa = String(a.perspective_slug ?? '');
    const sb = String(b.perspective_slug ?? '');
    return sa < sb ? -1 : sa > sb ? 1 : 0;
  })[0]!;
}

export interface PostNextCandidate {
  image_key: string;
  filename: string;
  date_taken: string;
  rating: number;
  peak_percentile: number;
  peak_perspective_slug: string;
  peak_perspective_display_name: string;
  is_signature: boolean;
  perspectives_covered: number;
  per_perspective: PeakPercentileItem['per_perspective'];
  reasons: string[];
  reason_codes: string[];
}

export interface PostNextPayload {
  candidates: PostNextCandidate[];
  total: number;
  meta: Record<string, unknown>;
  empty_state: string | null;
}

/**
 * Unposted, coverage-eligible catalog images with heuristic reasons (D-44–D-46).
 *
 * `sortByDate` only moves the tiebreaker; the corroboration-vetoed ranking
 * percentile stays the primary sort key.
 */
export function suggestWhatToPostNext(
  db: Db,
  opts: { limit: number; offset?: number; sortByDate?: 'newest' | 'oldest' | null },
): PostNextPayload {
  const sortByDate = opts.sortByDate ?? null;
  if (sortByDate !== null && sortByDate !== 'newest' && sortByDate !== 'oldest') {
    throw new RangeError("sort_by_date must be 'newest' or 'oldest'");
  }
  const offset = opts.offset ?? 0;

  // One index, three consumers: the score table is scanned once rather than
  // three times.
  const scan = buildCatalogScoreIndex(db);
  const { items, meta: peakMeta } = computeImagePeakPercentileScores(scan, {
    includeIneligible: false,
  });
  const crowned = new Set(
    computeSignatureStats(scan)
      .stats.filter((s) => s.crowned)
      .map((s) => s.perspective_slug),
  );

  const eligibleKeys = items.filter((i) => i.eligible).map((i) => String(i.image_key));
  const imgMeta = imageMetaMap(db, eligibleKeys);

  interface Candidate extends PeakPercentileItem {
    filename?: string;
    date_taken?: string;
    rating?: number;
    instagram_posted?: boolean;
    peak_perspective_slug: string;
    peak_perspective_display_name: string;
    is_signature: boolean;
  }

  const candidatesFull: Candidate[] = [];
  for (const i of items) {
    if (!i.eligible) continue;
    const k = String(i.image_key);
    const im = imgMeta.get(k);
    // Already posted: not a suggestion for what to post next.
    if (im?.instagram_posted) continue;
    const per = i.per_perspective ?? [];
    const peakLens = per.length > 0 ? peakPerspective(per) : {};
    const slug = String(peakLens.perspective_slug ?? '');
    candidatesFull.push({
      ...i,
      ...(im ?? {}),
      peak_perspective_slug: slug,
      peak_perspective_display_name: String(peakLens.display_name ?? slug),
      is_signature: slug ? crowned.has(slug) : false,
    });
  }

  sortByRanking(candidatesFull, sortByDate);

  // p90 of the *unposted eligible* population, not the whole catalog: the reason
  // code means "strong among what you have left to post".
  const rankings = candidatesFull.map((c) => Number(c.ranking_percentile));
  let p90 = 1;
  if (rankings.length > 0) {
    const sorted = [...rankings].sort((a, b) => a - b);
    const idx = Math.max(0, roundHalfEven(0.9 * (sorted.length - 1)));
    p90 = sorted[idx]!;
  }

  const suggestionsMeta = {
    weighting: peakMeta.weighting,
    ranking_key: peakMeta.ranking_key,
    corroboration_rule: peakMeta.corroboration_rule,
    min_perspectives_used: peakMeta.min_perspectives_used,
    coverage_rule: peakMeta.coverage_rule,
    high_score_rule:
      'reason code high_score_unposted when ranking_percentile >= p90 of eligible ' +
      "unposted images' corroboration-vetoed ranking percentiles.",
  };

  const total = candidatesFull.length;
  const lim = Math.max(0, opts.limit);
  const page = candidatesFull.slice(offset, offset + lim);

  const outCandidates: PostNextCandidate[] = page.map((cand) => {
    const reasons: string[] = [];
    const codes: string[] = [];
    const ranking = Number(cand.ranking_percentile);
    const lensName = String(cand.peak_perspective_display_name || '');

    if (ranking >= p90) {
      reasons.push(
        'Strong ranking percentile among scored, unposted catalog images ' +
          `(ranking_percentile=${ranking.toFixed(4)}).`,
      );
      codes.push('high_score_unposted');
    }

    if (lensName) {
      reasons.push(
        cand.is_signature
          ? `Peak lens is ${lensName}, one of your crowned signature techniques.`
          : `Peak lens is ${lensName}.`,
      );
    }

    // Only when no *code* was added — the lens sentence above is a reason but not a
    // reason code, so an image can carry it and still fall back to eligible_unposted.
    if (codes.length === 0) {
      reasons.push(
        'Unposted catalog image with sufficient perspective coverage; ' +
          'ranked by corroboration-vetoed ranking percentile among eligible candidates.',
      );
      codes.push('eligible_unposted');
    }

    return {
      image_key: cand.image_key,
      filename: cand.filename ?? '',
      date_taken: cand.date_taken ?? '',
      rating: cand.rating ?? 0,
      peak_percentile: cand.peak_percentile,
      peak_perspective_slug: cand.peak_perspective_slug ?? '',
      peak_perspective_display_name: cand.peak_perspective_display_name ?? '',
      is_signature: Boolean(cand.is_signature),
      perspectives_covered: cand.perspectives_covered,
      per_perspective: cand.per_perspective ?? [],
      reasons,
      reason_codes: codes,
    };
  });

  return {
    candidates: outCandidates,
    total,
    meta: suggestionsMeta,
    empty_state:
      outCandidates.length === 0
        ? 'No unposted catalog images meet perspective coverage with current scores.'
        : null,
  };
}
